repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,619 @@
+      // Add standard headers
+      request.addHeader(PowerApiHeader.USER.toString, token.username)
+      request.addHeader(OpenStackServiceHeader.USER_NAME.toString, token.username)
+      request.addHeader(OpenStackServiceHeader.USER_ID.toString, token.userId)
+      request.addHeader(OpenStackServiceHeader.TENANT_NAME.toString, token.tenantName)
+      request.addHeader(OpenStackServiceHeader.X_EXPIRATION.toString, token.expirationDate)
+      token.defaultRegion.foreach(request.addHeader(OpenStackServiceHeader.DEFAULT_REGION.toString, _))
+      token.contactId.foreach(request.addHeader(OpenStackServiceHeader.CONTACT_ID.toString, _))
+      token.impersonatorId.foreach(request.addHeader(OpenStackServiceHeader.IMPERSONATOR_ID.toString, _))
+      token.impersonatorName.foreach(request.addHeader(OpenStackServiceHeader.IMPERSONATOR_NAME.toString, _))
+
+      // Construct and add the tenant id header
+      val tenantsToPass = buildTenantHeader(token.defaultTenantId, token.tenantIds, tenantFromUri)
+      request.addHeader(OpenStackServiceHeader.TENANT_ID.toString, tenantsToPass.mkString(","))
+
+      // If configured, add roles header
+      if (config.getIdentityService.isSetRolesInHeader) {
+        request.addHeader(OpenStackServiceHeader.ROLES.toString, token.roles.mkString(","))
+      }
+
+      // If present, add the tenant from the URI as part of the Proxy header, otherwise use the default tenant id
+      tenantFromUri match {
+        case Some(uriTenant) =>
+          request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION.toString, s"$X_AUTH_PROXY $uriTenant")
+        case None =>
+          request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION.toString, s"$X_AUTH_PROXY ${token.defaultTenantId}")
+      }
+    }
+
+    def addCatalogHeader(maybeEndpoints: => Try[EndpointsData]): Try[Unit.type] = {
+      if (config.getIdentityService.isSetCatalogInHeader) {
+        maybeEndpoints map { endpoints =>
+          request.addHeader(PowerApiHeader.X_CATALOG.toString, Base64.encodeBase64String(endpoints.json.getBytes))
+          Unit
+        }
+      } else {
+        Success(Unit)
+      }
+    }
+
+    def addGroupsHeader(maybeGroups: => Try[Vector[String]]): Try[Unit.type] = {
+      if (config.getIdentityService.isSetGroupsInHeader) {
+        maybeGroups map { groups =>
+          if (groups.nonEmpty) {
+            request.addHeader(PowerApiHeader.GROUPS, groups.mkString(","))
+          }
+          Unit
+        }
+      } else {
+        Success(Unit)
+      }
+    }
+
+    def addIdentityStatusHeader(confirmed: Boolean): Unit = {
+      if (Option(config.getDelegating).isDefined) {
+        if (confirmed) request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.Confirmed.toString)
+        else request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.Indeterminate.toString)
+      }
+    }
+
+    def getIdentityServiceUri(identityServiceUri: String): String = {
```

Comment:
I also kind of dislike having this function as the point-of-failure if a uri is the empty string. Perhaps we should use lastOption rather than last, and do nothing if lastOption returns None.